### DISCUSSION TITLE
Multiple vulnerabilities in Airflow dependencies

### DISCUSSION THREAD

**Discussion by vahedq** [apache/airflow#24055] [General] [answered]
We're relying heavily on Airflow, and have noticed that the latest constraint file below results in a number of high and critical vulnerabilities. 
```
https://raw.githubusercontent.com/apache/airflow/constraints-2.3.1/constraints-3.9.txt
```

Is there a plan to update these dependencies to more recent versions to resolve vulnerabilities? If not, we're wondering how everyone is handling/fixing these vulnerabilities? We tried pinning some to newer versions, but that breaks airflow. Any help is appreciated.

```
CVE-2021-33026 | Flask-Caching | CRITICAL 
GHSA-h8pj-cxx2-jfg2 | httpx | CRITICAL | 
CVE-2021-41945 | httpx | CRITICAL |
IN1-PYTHON-FLASKCACHING-1292339 | Flask-Caching | HIGH |
GHSA-ffqj-6fqr-9h24 | PyJWT | HIGH 
CVE-2022-29217 | PyJWT | MEDIUM | 
IN1-PYTHON-WTFORMS-40581 | WTForms | MEDIUM
IN1-PYTHON-PYJWT-2840625 | PyJWT | MEDIUM 
```

**Comment by potiuk** [comment, marked as answer]
We will likely upgrade Flask App Builder to 4.0.0 soon-ish which would solve some of those - there is an issue #22397 about it. Hopefully someone will look at it soon.

Any contriburions are welcome to update it - so if you could contribute it back - anyone can make a PR to update to FAB an test it.  Airflow is created by more than 2000 contributors, so you could definitely contribute it back. Since you are vitaly interested - maybe your company could take a look at it  @vahedq ?

**Reply by vahedq** [reply]
Thanks @potiuk we will try our best to help out.

**Comment by dovregubben** [comment]
In Airflow 2.10.5 we still see CVE-2021-33026, due to the constraint of Flask-Caching being pinned to 2.3.0 (see https://raw.githubusercontent.com/apache/airflow/constraints-2.10.5/constraints-3.12.txt).
Could anybody elaborate on why this CVE has not been taken into account yet, or if the vulnerabilty has been classified differently for Apache Airflow?

**Reply by potiuk** [reply]
Because it's you to take care about your deployment security when new versions of dependencies are releaed. The CVE and new release of the dependency happened yesterday. And yes you should upgrade as soon as possible if you are worrried, it's your responsibility to keep your deployment secure.

As explained here https://airflow.apache.org/docs/docker-stack/index.html#fixing-images-at-release-time - we are fixing our images using constraint at release time, and it's up to deployment managers to update new dependencies between releases. We even explain how to do it (follow links there).

Also you are completely mistaken about constraints. They are not pinning the dependencies. Constraints are exclusively done in order to allow reproducible installs (**reproducible** not **keep latest security dependencies fixed**). Except cases where airflow stops installing because of external dependency got some really weird problems, we are not updating constraints for released versions - and leave it up to the users to upgrade airflow dependencies with security fixes **without** using constraints. You can read more here: https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#why-we-need-constraints .

Constraints are "constraints" and not "requirements" - precisely because they allow you to upgrade dependencies (after initial installation that is reproducible) with newly released dependencies, where it is possible. You should not expect constraints for relased version of Airlfow to be updated ever, it happend maybe 3 times in the past when setuptools and others similar weird issues happend that prevented installing airlfow with newer `pip` or newer `setuptools` installed. 

As you can see our "main" version of constraints - which wil be used when we release next version of Airlfow already bumped Flask-Caching to 2.3.1 https://github.com/apache/airflow/commit/2a4362346132951dccacf96e2ef75b62a1174b44 pretty much immediately after it's been released (we have automation of those upgrades), so I don't think there should be any blocker for you to upgrade.

So summarizing: yes good call about the CVE (It's cool that you are watching them and reacting so quickly), and yes it's totally on you to upgrade and build your own process of updating the dependencies in the way that such upgrades are accounted for - and you have good deal of material, docs, instructions and examples to do so.

You can also watch my talk from the airflow summit where I explain how constraints, incrementdal upgrades work and how you should approach this. Might be useful for you to build your process https://www.youtube.com/watch?v=zPjIQjjjyHI

Also Airlfow Summit is coming in October to Seattle. Good way to learn more about airlfow. 

I hope that clarifies it.

**Reply by potiuk** [reply]
BTW. Your CVE is likely completely mixed up. The CVE you pointed out is very old and affects Flask-Caching `<= 1.10.1` and it is very old, so you likely made wrong reference.

**Reply by dovregubben** [reply]
I very much appreciate your explanation, the links and the recording of your talk, all of that really helped me a lot in several ways.
I was indeed mislead by the concept of constraints. It's very helpful to understand that those are mainly shipped for reproducibilty and that we can simply bump or add specific packages in a second step (after setting up `apache-airflow`).
With regard to the CVE I mentioned, I literally meant CVE-2021-33026. A critical finding concerning that CVE was raised to us by the AWS Inspector, that's why I was asking about it in the first place. However your last comment made me investigate a bit more, and now I understand that this might rather be a false positive warning, because both from the description of the vulnerability and also the GitHub Advisory Database I can see that it only affects Flask-Caching `<= 1.10.1`.
So it appears to me that the AWS Inspector doesn't pick up the information right from NVD.
I also followed up on your discussion regarding this CVE in https://github.com/apache/airflow/issues/16541.
